homeassistant/components/light/lifx.py

```diff
@@ -6,124 +6,97 @@
-    def find_bulb(self, ipaddr):
-        """Search for bulbs."""
-        bulb = None
-        for device in self._devices:
-            if device.ipaddr == ipaddr:
-                bulb = device
-                break
-        return bulb
+    """Representation of all known LIFX entities."""
 
-    def on_device(self, ipaddr, name, power, hue, sat, bri, kel):
+    def __init__(self, hass, async_add_devices):
         """Initialize the light."""
-        bulb = self.find_bulb(ipaddr)
-
-        if bulb is None:
-            _LOGGER.debug("new bulb %s %s %d %d %d %d %d",
-                          ipaddr, name, power, hue, sat, bri, kel)
-            bulb = LIFXLight(
-                self._liffylights, ipaddr, name, power, hue, sat, bri, kel)
-            self._devices.append(bulb)
-            self._add_devices_callback([bulb])
+        self.entities = {}
+        self.hass = hass
+        self.async_add_devices = async_add_devices
+
+    def register(self, device):
+        """Callback for newly detected bulb."""
+        if device.mac_addr in self.entities:
+            entity = self.entities[device.mac_addr]
+            _LOGGER.debug("%s register AGAIN", entity.ipaddr)
+            entity.available = True
+            entity.schedule_update_ha_state()
         else:
-            _LOGGER.debug("update bulb %s %s %d %d %d %d %d",
-                          ipaddr, name, power, hue, sat, bri, kel)
-            bulb.set_power(power)
-            bulb.set_color(hue, sat, bri, kel)
-            bulb.schedule_update_ha_state()
+            _LOGGER.debug("%s register NEW", device.ip_addr)
+            device.get_color(self.ready)
 
-    def on_color(self, ipaddr, hue, sat, bri, kel):
-        """Initialize the light."""
-        bulb = self.find_bulb(ipaddr)
-
-        if bulb is not None:
-            bulb.set_color(hue, sat, bri, kel)
-            bulb.schedule_update_ha_state()
-
-    def on_power(self, ipaddr, power):
-        """Initialize the light."""
-        bulb = self.find_bulb(ipaddr)
-
-        if bulb is not None:
-            bulb.set_power(power)
-            bulb.schedule_update_ha_state()
-
-    # pylint: disable=unused-argument
-    def poll(self, now):
-        """Polling for the light."""
-        self.probe()
+    def ready(self, device, msg):
+        """Callback that adds the device once all data is retrieved."""
+        entity = LIFXLight(device)
+        _LOGGER.debug("%s register READY", entity.ipaddr)
+        self.entities[device.mac_addr] = entity
+        self.hass.async_add_job(self.async_add_devices, [entity])
 
-    def probe(self, address=None):
-        """Probe the light."""
-        self._liffylights.probe(address)
+    def unregister(self, device):
+        """Callback for disappearing bulb."""
+        entity = self.entities[device.mac_addr]
+        _LOGGER.debug("%s unregister", entity.ipaddr)
+        entity.available = False
+        entity.updated_event.set()
+        entity.schedule_update_ha_state()
```

Comment:
You are already in the event loop. Please use `self.hass.async_add_job(entity.async_update_ha_state())`